File: maximum_twin_sum_of_a_linked_list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {number}
 */
var pairSum = function(head) {
    let slow = head;
    let fast = head;

    while (fast && fast.next) {
        slow = slow.next;
        fast = fast.next.next;
    }

    let second = null;

    while (slow) {
        let temp = slow.next;
        slow.next = second;
        second = slow;
        slow = temp;
    }

    let res = 0;
    while (second) {
        res = Math.max(res, head.val + second.val);
        head = head.next;
        second = second.next;
    }

    return res;    
};

File: maximum_twin_sum_of_a_linked_list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def pairSum(self, head: Optional[ListNode]) -> int:

        slow = head
        fast = head

        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
        
        second = None

        while slow:
            temp = slow.next
            slow.next = second
            second = slow
            slow = temp
        
        res = 0
        while second:
            res = max(res, head.val + second.val)
            head = head.next
            second = second.next
        
        return res
